src/libraries/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/Http2Connection.cs

```diff
@@ -766,111 +760,184 @@ private void ProcessGoAwayFrame(FrameHeader frameHeader)
         }
 
         internal Task FlushAsync(CancellationToken cancellationToken) =>
-            PerformWriteAsync(0, 0, (_, __) => FlushTiming.Now, cancellationToken);
-
-        /// <summary>Performs a write operation serialized via the <see cref="_writerLock"/>.</summary>
-        /// <param name="writeBytes">The number of bytes to be written.</param>
-        /// <param name="state">The state to pass through to the callbacks.</param>
-        /// <param name="lockedAction">The action to be invoked while the writer lock is held and that actually writes the data to the provided buffer.</param>
-        /// <param name="cancellationToken">The cancellation token to use while waiting.</param>
-        private async Task PerformWriteAsync<T>(int writeBytes, T state, Func<T, Memory<byte>, FlushTiming> lockedAction, CancellationToken cancellationToken = default)
+            PerformWriteAsync(0, 0, (_, __) => true, cancellationToken);
+
+        private abstract class WriteQueueEntry : TaskCompletionSource
```

Comment:
Doesn't need to be part of this PR, but given we're now allocating this on every write, it'd be really good to look at changing the base type to be one that implements IValueTaskSource and making this reusable.

Reply:
Agreed, though it gets a little weird because we have a bunch of different write operations with different kinds of state.

Reply:
There's a tradeoff to be made here. This is similar to what pipelines does. You don't wait on individual writes but you can have backpressure based on how much outstanding buffer is unflushed/unconsumed. That removes the per operation overhead and allocations.

Do you care about cancelling individual writes or do you care about stopping the overall writing operation?